### ISSUE TITLE
Add a function to remove optional attributes with null value to Node

### ISSUE THREAD

**Issue by LourensVeen** [yatiml/yatiml#8]
Optional attributes may be omitted on input. On output they are always written though, even if they have a None value. I don't think we want to always have YAtiML strip these attributes, but it should be easy to add in yatiml_sweeten(). To enable that a Node.remove_optional_null_attributes() method should be added.

Actually, this should only be done if the default value for the attribute is None. If it has a different default, then doing this would break round-tripping. Although I guess in that case round-tripping would be broken too if we write it as a null, because that would cause the object to have the default value. On the other hand, how would you construct such an object? You'd have to construct it first, which would set the non-None default value, and then overwrite that with a None. Needs a bit more thought.

**Comment by LourensVeen** [comment]
So more generally, should attributes with a value that matches their default value be omitted from the output automatically? Or should there be a Node.remove_attributes_with_default_value()? Or a Node.keep_default_valued_attributes()? Implementing that is tricky, because Node doesn't have access to the class we're trying to serialise or construct, so how does it know the default value?

Right now though, the user has to manually remove the attribute if the value matches the default, and so ends up with two places where that value is defined, once in __init__ and once in the sweeten or savorize function.

Hmm, since sweeten and savorize are class methods, maybe we can have the user pass the cls parameter?

**Comment by LourensVeen** [comment]
Reading back the top comment, is it really the case that an attribute set explicitly to None will result in the object getting the default value? It would be more logical if only a missing attribute would result in the default, and an explicit null would result in None being passed to the constructor on loading. Should check that.

**Comment by LourensVeen** [comment]
There are two different things here, whether a parameter is optional, and whether is has a default. If it's optional, you're allowed to pass None, but not allowed to omit the argument. If it's defaulted, you're allowed to omit, but None is not per se a valid argument.

So to make this work, optional is irrelevant; what matters is whether there's a default value, and whether the current value of the attribute matches the default of the corresponding parameter. To know that, we need access to the current mapping (which we have in Node) and to the corresponding class (which we can get by having the user pass the cls argument).

There's one more problem: if you have a parameter that expects an object, then setting a default is tricky because each instance will get a reference to the same object. If the parameter is, say, for a list, and you store that list as an attribute and later modify it, then you'll end up with all instances having those modifications. A [common work-around](https://docs.python-guide.org/writing/gotchas/) is to make the attribute optional and pass None as the default, then check whether it's None in the `__init__` and replace with a fresh empty list. Of course, YAtiML doesn't know that, and won't omit the attribute if it's an empty list.

Arguably, in the face of typed Python, that work-around is no longer so nice, because you're changing the type of the parameter. Perhaps it would be better to set an empty list as the default, then in `__init__` check for that and copy it or replace it with a new list. For example:

```
T = TypeVar('T')

def append_to(element: T, to: List[T]=[]):
    if not to:
        to = list()
    to.append(element)
    return to
```
That'll probably still set off your linter though, although you could disable that check.

Anyway, I guess we could add a `_yatiml_defaults: Dict[str, Any]` class attribute which overrides the type annotation? Then the parameter can be optional and defaulted to None, but if you set `_yatiml_defaults = {'p': []}` then YAtiML will use that instead?

**Comment by LourensVeen** [comment]
Released with 0.6.0.
